src/main/scala/org/bblfsh/client/libuast/alloclist.c

```diff
@@ -0,0 +1,38 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "alloclist.h"
+
+#include <jni.h>
+
+#include <stdlib.h>
+
+void initAllocList(AllocList *l, size_t initialSize) {
+  l->list = (jobject **)malloc(initialSize * sizeof(jobject*));
+  l->used = 0;
+  l->size = initialSize;
+}
+
+void trackAllocatedJObject(AllocList *l, jobject *obj) {
+  if (l->used == l->size) {
+    l->size *= 2;
+    l->list = (jobject **)realloc(l->list, l->size * sizeof(jobject*));
+  }
+  l->list[l->used++] = obj;
+}
+
+void freeAllocList(AllocList *l) {
+  int i;
+  for (i=0; i < l->used; i++) {
+    free(l->list[i]);
+  }
+
+  free(l->list);
+  l->list = NULL;
+  l->used = l->size = 0;
```

Comment:
If you call `trackAllocatedJObject` after this, you have memory corruption. A sanity check in `trackAllocationJObject` may be a good idea.

Reply:
No corruption, it would read a NULL pointer which would cause a SIGSEGV. Another option would be to detect that the vector is not initialized on `trackAllocatedJObject` and do the initialization there. What do you think?

Reply:
PS: check the last commit, I just did that (removed `initAllocVector` from the public interface and made `trackAllocatedObject` call it if needed).

Reply:
No, it would not read a NULL pointer. It would check that `size` is equal to `used` and then would call `realloc` with NULL and 0 `size`, which may or may not return NULL, then that pointer will be accessed... so it doesn't sound safe at all :)

This should be fixed with your last commit though (but I wouldn't make the vector struct public, though.)

Reply:
I checked that realloc with the NULL argument produces a sigsev in that case.

Anyway... it doesn't work :-| Since the freeing is at the end of `filter()` call and `filter` returns data, freeing part of that data produces a crash later if you try to use it (from Scala), so I'll take a look at how to hook with JNI into the Scala object destructor to call the free there.

Reply:
Luckily the problem wasn't that complex but an `extern` declaration of the vector as a pointer (extern AllocVector *allocVector)with the real declaration as non pointer... surprising that `gcc` didn't even issue a warning, but such is life. That was fixed following your advice of making the vector private to the file :+1: